cl/phase1/stages/stages.go

```diff
@@ -2,89 +2,271 @@ package stages
+					"currentSlot", currentSlot,
+				)
+				// now we download the missing blocks
+
+				type resp struct {
+					blocks []*cltypes.SignedBeaconBlock
+				}
+				chans := make([]chan resp, 0, totalEpochs)
+				ctx, cn := context.WithCancel(ctx)
+				egg, ctx := errgroup.WithContext(ctx)
+				egg.SetLimit(8)
+				defer cn()
+				for i := seenEpoch; i <= targetEpoch; i = i + 1 {
+					ii := i
+					o := make(chan resp, 0)
+					chans = append(chans, o)
+					egg.Go(func() error {
+						logger.Debug("request epoch from reqresp", "epoch", ii)
+						blocks, err := cfg.source.GetRange(ctx, ii*cfg.beaconCfg.SlotsPerEpoch, cfg.beaconCfg.SlotsPerEpoch)
+						if err != nil {
+							return err
+						}
+						logger.Debug("got epoch from reqresp", "epoch", ii)
+						o <- resp{blocks}
+						return nil
+					})
+				}
+				errchan := make(chan error)
+				go func() {
+					defer func() {
+						close(errchan)
+					}()
+					for _, v := range chans {
+						select {
+						case <-ctx.Done():
+							return
+						default:
+						}
+						epochResp := <-v
+						for _, block := range epochResp.blocks {
+							if block.Block.Slot <= seenSlot {
+								continue
+							}
+							if err := cfg.forkChoice.OnBlock(block, false, true); err != nil {
+								log.Warn("fail to process block", "reason", err, "slot", block.Block.Slot)
+								err2 := cfg.source.PurgeRange(context.Background(), block.Block.Slot, totalEpochs)
+								if err2 != nil {
+									log.Error("failed to purge range", "err", err2)
+								}
+								errchan <- err
+								return
+							}
+							saveBlock(block)
+						}
+					}
+				}()
+				err := egg.Wait()
+				if err != nil {
+					return err
+				}
+				return <-errchan
 			},
 			Unwind: func(firstCycle bool, u *stagedsync.UnwindState, s *stagedsync.StageState, tx kv.RwTx, logger log.Logger) error {
 				return nil
 			},
 		},
+
 		{
-			ID:          stages.BeaconBlocks,
-			Description: "Download beacon blocks forward.",
+			ID: "CatchUpBlocks",
+			Description: `this stage runs if the current node is not at head, otherwise it moves on.
+			`,
 			Forward: func(firstCycle bool, badBlockUnwind bool, s *stagedsync.StageState, u stagedsync.Unwinder, tx kv.RwTx, logger log.Logger) error {
-				return SpawnStageForkChoice(forkchoice, s, tx, ctx)
+				seenSlot := cfg.forkChoice.HighestSeen()
+				targetSlot := utils.GetCurrentSlot(cfg.genesisCfg.GenesisTime, cfg.beaconCfg.SecondsPerSlot)
+
+				seenEpoch := seenSlot / cfg.beaconCfg.SlotsPerEpoch
+				targetEpoch := utils.GetCurrentEpoch(cfg.genesisCfg.GenesisTime, cfg.beaconCfg.SecondsPerSlot, cfg.beaconCfg.SlotsPerEpoch) - 1
+				if seenEpoch < targetEpoch {
+					return nil
+				}
+				if seenSlot >= targetSlot {
+					return nil
+				}
+
+				// wait for three slots... should be plenty enough time
+				ctx, cn := context.WithTimeout(ctx, time.Duration(cfg.beaconCfg.SecondsPerSlot)*time.Second*3)
+				defer cn()
+
+				go func() {
+					cfg := forkchoice
+					for {
```

Comment:
you probably so if the chain is at block 36, you would arrive at block 32 and get stuck because you miss 4 blocks. you should use reqresp here as well